osrf/vrx
# gymkhana scoring plugin

## To test
Launch the task:
```
ros2 launch vrx_gz competition.launch.py world:=gymkhana_task
```
In a separate terminal, launch the teleop:
```
ros2 launch vrx_gz usv_joy_teleop.py
```
In a third terminal, echo the `/vrx/task/info` topic:
```
ros2 topic echo /vrx/task/info
```

### Test Navigation Channel
* Switch back to the first terminal.
* Wait for the task to enter `onrunning` state.
* Use the gamepad to drive the WAMV through the gates in the course.
* Check that a "New gate crossed!" message is printed to terminal each time a gate is crossed.
* Check that the navigation scoring plugin exits after all 3 gates are crossed, but not the simulation. You should see a message that says:
  ```
  [Err] [ScoringPlugin.cc:330] VRX_EXIT_ON_COMPLETION and <per_plugin_exit_on_completion> both not set, will not shutdown on ScoringPlugin::Exit()
  ```
* Check that the score is reasonable. If you went through relatively quickly, your score should be about 88 when you exit the channel.
*  Check that the score matches the mean pose error.
  ```
  ros2 topic echo /vrx/gymkhana_blackbox/mean_pose_error
  ```
  It should give you the same value.
* Check that the pinger output is reasonable:
  ```
  ros2 topic echo /wamv/pingers/pinger/range_bearing
  ```
  If your wamv is at the end of the navigation channel, the range should be about 86.
* Check that the range approximately matches the pose error:
  ```
  ros2 topic echo /vrx/gymkhana_blackbox/pose_error
  ```
  It should also be about 86. If you wait long enough, all 4 of the above values will converge.
* Pause the simulation.

### Test Stationkeeping plugin
With the simulation still paused, use the GUI to teleport the WAMV to the pinger position:
```
X = -485
Y= 295.5
```
* Unpause. 
* Check `/wamv/pingers/pinger/range_bearing` and `/vrx/gymkhana_blackbox/pose_error`. The range and the pose error should be similar and near zero.
* Check `/vrx/task/info` and `/vrx/gymkhana_blackbox/mean_pose_error`. These should be the same and should be steadily decreasing.

## Known Limitations
* WAMV is not locking properly. See issue #545
* Collisions don't seem to be detected yet. I will address this in a separate PR.
